src/index.d.ts

```diff
@@ -0,0 +1,721 @@
+ * Anything else will be converted to a string and then
+ * turned into a buffer.
+ */
+interface TidyBufferStatic {
+  (document: string | Buffer, options: Generated.OptionDict,
+    callback: TidyCallback): void
+
+  (document: string | Buffer, callback: TidyCallback): void
+}
+
+export class TidyOption {
+  // creation of TidyOption is not exposed
+  private constructor()
+  readonly name: string
+  readonly type: "boolean" | "integer" | "doctype" | "string"
+  readonly readOnly: boolean
+  readonly pickList: string[]
+  readonly id: number
+  readonly category: string
+  toString(): string
+}
+
+/**
+ * TidyDoc is the central object for dealing with the library at a low level
+ *
+ * Such an object will hold a configuration and be able to process one input
+ * file at a time, while multiple such objects can deal with multiple inputs
+ * simultaneously using an independent configuration for each of them.
+ */
+interface TidyDoc extends Generated.TidyDocOption {
+  // Sync calls
+  cleanAndRepairSync(): string
+  parseBufferSync(document: Buffer): string
+  runDiagnosticsSync(): string
+  saveBufferSync(): Buffer
+  // getErrorLog(): string // is not needed: other calls already return log
+
+  // Async calls
+  cleanAndRepair(callback: TidyCallback): void
+  parseBuffer(document: Buffer, callback: TidyCallback): void
+  runDiagnostics(callback: TidyCallback): void
+  saveBuffer(callback: TidyCallback): void
+  tidyBuffer(buf: Buffer, callback: TidyCallback): void
+
+  // batch set/get of options
+  options: Generated.OptionDict
+  // Methods that return TidyOption object
+  getOptionList(): TidyOption[]
+  getOption(option: TidyOption): TidyOption
+  getOption(optionId: number): TidyOption
+  getOption(optionName: string): TidyOption
+  /**
+   * FIXME: does this return the list of related options??
+   */
+  optGetDocLinksList(optionId: number): TidyOption[]
+  optGetDocLinksList(optionName: string): TidyOption[]
+  /* FIXME: what does this do? */
+  optGetCurrPick(key: any): any
+}
+
+/**
+ * Constructor
+ * (Can be used with `new` or normal call)
+ */
+interface TidyDocConstructor {
+  new (): TidyDoc
+  (): TidyDoc
+}
+
+/**
+ * Elements of the compat namespace offer compatibility with other
+ * libtidy bindings for node.
+ */
+interface TidyCompat {
+  libtidy: {
+    tidy: {
+      (text: string, callback: (err: any, html: string) => void): void
+      (text: string, options: any, callback: (err: any, html: string) => void): void
+    }
+  }
+}
+
+// START-GENERATED-NAMESPACE
+// The rest of this file is generated.
+/**
+* Type for libtidy options
+* @generated with /util/gen-typescript-decl.ts
+*/
+declare namespace Generated {
+  interface TidyDocOption {
+    /**
+    * PrettyPrint / indent-spaces (integer)
+    */
+    optGet(key: "indent-spaces" | "indent_spaces" | "IndentSpaces"): number;
+    optSet(key: "indent-spaces" | "indent_spaces" | "IndentSpaces", value: number): void;
```

Comment:
Does this level of detail for the getters and setters really make sense? It does sound reasonable if you really are configuring these things manually, using string literals as keys. But those applications are easier to write using the options dictionary notation, I think. What happens if the configuration comes from a JSON file or similar? Is there some easy solution to make that work, too, or would that mean you'd have to type-cast the key from the JSON file to the enum of all the alternatives you have configured here? Would it make sense to add a catch-all definition up front? For the getter, that would be easy: `optGet(key: string): number | string | boolean` I think. For the setter that would become `optSet(key: string, value: number | string | boolean)` which would defy all the special cases, as this one case would capture anything not captured by the special cases, thus breaking the type checking there. So should we only do the specific getters, and the generic getter and setter? I'd like to see a generic get and set (using separately typed variables for key and value) demonstrated in the unit test.